Add documentation about working in virtualenvs to the Matplotlib faq

jenshnielsen opened the pull request:
The writing could probably still use some work.

tacaswell commented on doc/faq/virtualenv_faq.rst:
```diff
@@ -0,0 +1,104 @@
+============
+
+When running :mod:`matplotlib` in a virtual environment you may discover a
+few issues. :mod:`matplotlib` itself works problemless in virtual envirnments.
```
"mpl itself has no issue with virtual environments."  ?

tacaswell commented on doc/faq/virtualenv_faq.rst:
```diff
@@ -0,0 +1,104 @@
+Most GUI frameworks are not pip installable. This makes it tricky to install
+them within a virtual environment. This problem does not exist if you use Conda
+environments where you can install all Conda supported GUI frameworks directly
+into tge environment. In regular virtualenv environment various workarounds
```
tge -> the

tacaswell commented on doc/faq/virtualenv_faq.rst:
```diff
@@ -0,0 +1,104 @@
+<https://github.com/pypa/virtualenv/issues/54>`__ and `here
+<https://github.com/pypa/virtualenv/issues/609>`__
+
+Until this is fixed a work around is needed. The best knows work around,
```
knows -> known

tacaswell commented:
attn @mdehoon

efiring commented on doc/faq/virtualenv_faq.rst:
```diff
@@ -0,0 +1,104 @@
+
+When running :mod:`matplotlib` in a virtual environment you may discover a
+few issues. :mod:`matplotlib` itself works problemless in virtual envirnments.
+However, the GUI frameworks that :mod:`matplotlib` depends uses for interactive
```
delete "depends"

efiring commented on doc/faq/virtualenv_faq.rst:
```diff
@@ -0,0 +1,104 @@
+backend? <what-is-a-backend>`.
+
+If you only use the ``IPython/Jupyter`` inline and ``notebook/nbagg`` backends
+and non interactive backends you can should not have any isseus and can ignore
```
delete first "can"; typo, "issues"

efiring commented on doc/faq/virtualenv_faq.rst:
```diff
@@ -0,0 +1,104 @@
+However, this breaks the isolation between the virtual environment and the
+system install. If you use `virtualenvwrapper
+<https://virtualenvwrapper.readthedocs.org/>`_  this can be toggled with the
+toggleglobalsitepackages command.
```
format: double backquotes for `toggleglobalsitepackages` to make it stand out?

efiring commented on doc/faq/virtualenv_faq.rst:
```diff
@@ -0,0 +1,104 @@
+Alternatively you can manually symlink the GUI frameworks into the environment.
+I.e. to use PyQt5 you should symlink ``PyQt5`` and ``sip`` from your system
+site packages directory into the environment taking care that the environment
+and systemwide install uses the same python version.
```
insert "the" before "systemwide"; "uses" -> "use"

efiring commented on doc/faq/virtualenv_faq.rst:
```diff
@@ -0,0 +1,104 @@
+OSX
+===
+
+On OSX two different types of Python Builds exists. A regular build and a
```
-> "...exist: a regular..."

efiring commented on doc/faq/virtualenv_faq.rst:
```diff
@@ -0,0 +1,104 @@
+On OSX two different types of Python Builds exists. A regular build and a
+framework build. In order to interact correctly with OSX through some
+gui frameworks you need a framework build of Python.
+At the time of writing the ``macosx``, ``WX`` and ``WXAgg`` backends requires a
```
requires -> require

efiring commented on doc/faq/virtualenv_faq.rst:
```diff
@@ -0,0 +1,104 @@
+Until this is fixed a work around is needed. The best knows work around,
+borrowed  from the `WX wiki
+<http://wiki.wxpython.org/wxPythonVirtualenvOnMac>`_, is to  use the non
+virualenv python along with the PYTHONHOME environmental variable.  This can be
```
typo: virtualenv

efiring commented on doc/faq/virtualenv_faq.rst:
```diff
@@ -0,0 +1,104 @@
+<http://wiki.wxpython.org/wxPythonVirtualenvOnMac>`_, is to  use the non
+virualenv python along with the PYTHONHOME environmental variable.  This can be
+implemented in a script as below. To use this modify ``PYVER`` and
+``PATHTOPYTHON`` and put the script in the virtualenvs bin directory i.e.
```
"virtualenvs" -> "virtualenv" or "virtualenv's"

efiring commented on doc/faq/virtualenv_faq.rst:
```diff
@@ -0,0 +1,104 @@
+  exec $PYTHON "$@"
+
+
+With this in place you can run ``frameworkpython`` to get a interactive
```
"a" -> "an"

jenshnielsen commented:
@tacaswell and @efiring Thanks for the prof reading. I have corrected the typos that you spotted. I am also interested in if people thinks this is understandable from a technical point of view? Given that we have added the check to the OSX backend that makes it error out inside a virtual env I think it is very important that we add some docs about it before the release which are understandable. We have already had a few bug reports about this in the rc's

matthew-brett commented on doc/faq/virtualenv_faq.rst:
```diff
@@ -0,0 +1,105 @@
+<https://github.com/pypa/virtualenv/issues/54>`__ and `here
+<https://github.com/pypa/virtualenv/issues/609>`__
+
+Until this is fixed a work around is needed. The best known work around,
```
"workaround" should be one word : e.g. https://en.wiktionary.org/wiki/workaround

matthew-brett commented on doc/faq/virtualenv_faq.rst:
```diff
@@ -0,0 +1,105 @@
+Until this is fixed a work around is needed. The best known work around,
+borrowed  from the `WX wiki
+<http://wiki.wxpython.org/wxPythonVirtualenvOnMac>`_, is to  use the non
+virualenv python along with the PYTHONHOME environmental variable.  This can be
```
typo "virualenv"

efiring commented:
I agree that this is needed, and I think it is reasonably understandable.  Some writing tweaks could still be made.  Regarding the more important aspect, the substance: I'm a little concerned about advocating `--system-site-packages` as a viable option.  My impression is that this is a route to hard-to-debug trouble.  This is just an impression based on thrashing around with virtualenv a few months ago, after which I gave up on `--system-site-packages`, and concluded that virtualenv was not worth the fuss for any mpl gui other than tkagg.

efiring commented on doc/faq/virtualenv_faq.rst:
```diff
@@ -0,0 +1,105 @@
+some familiarity with the Matplotlib backends as found in :ref:`What is a
+backend? <what-is-a-backend>`.
+
+If you only use the ``IPython/Jupyter`` inline and ``notebook/nbagg`` backends
```
The `inline` also needs the backticks, and `notebook/nbagg` might be `notebook` or (alias) `nbagg`.  Correct?  Otherwise, I have a little trouble parsing this, and realizing that it is very specific; just using the Jupyter notebook, for example, is not enough.  This is in contrast to ordinary gui operations, in which using IPython or the notebook tends to solve all problems, regardless of backend.  (Well, most, anyway.)

jenshnielsen replied:
Yes

jenshnielsen commented:
I agree about the `--system-site-packages` I was trying to hint something about that in the original comment but have made it more clear now. I still think we need to comment on it because it's mentioned in a lot of docs/posts on the internet.

matthew-brett commented:
It is completely necessary to error out in a virtualenv for MacOSX backend?  Presumably it would in fact work - as it did in 1.4.3?

jenshnielsen commented:
That was the decision made by @mdehoon and I tend to agree with it. The issue is that the figure window cannot be bought to the foreground and seemingly random parts of the interaction with the figure doesn't work resulting in puzzling bug reports which are hard to debug

matthew-brett commented:
Oh dear - that will be a problem for me, I use the OSX backend a lot.

How about a non-default rcparam to enable the OSX backend in virtualenvs?

matthew-brett commented:
Does the OSX backend work better in conda environments?

jenshnielsen commented:
AFAIK Conda envs have the same issue. I don't think a RC param to enable something which is fundamentally broken is a good idea. I would suggest to use the `PYTHONHOME` workaround

jenshnielsen commented:
We can consider changing it to a warning if people prefer that?

mdboom commented on doc/faq/virtualenv_faq.rst:
```diff
@@ -0,0 +1,106 @@
+Introduction
+============
+
+When running :mod:`matplotlib` in a virtual environment you may discover a
```
We should probably start with links to what we're talking about.

https://virtualenv.pypa.io/en/latest/

or, in Python 3.3 and later:

https://docs.python.org/3/library/venv.html

mdboom replied:
Can you get this?

jenshnielsen replied:
Sorry I must have missed this. I will fix if asap

mdboom commented on doc/faq/virtualenv_faq.rst:
```diff
@@ -0,0 +1,106 @@
+into the environment. In regular virtualenv environment various workarounds
+exist. Some of these are given here:
+
+The ``TKAgg`` backend doesn't require any external dependencies and is normally
```
Maybe make these separate bullet points, one for TkAgg and one for PySide, since I think if you were skimming this for info about a particular backend, you may miss PySide.

mdboom commented on doc/faq/virtualenv_faq.rst:
```diff
@@ -0,0 +1,106 @@
+One often suggested solution is to use the ``--system-site-packages`` option
+to virtualenv when creating an environment. This adds all system wide packages
+to the virtual environment. However, this breaks the isolation between the
+virtual environment and the system install. Among other issues it result in
```
result -> results

mdboom commented on doc/faq/virtualenv_faq.rst:
```diff
@@ -0,0 +1,106 @@
+If you use `virtualenvwrapper <https://virtualenvwrapper.readthedocs.org/>`_
+this can be toggled with the ``toggleglobalsitepackages`` command.
+
+Alternatively you can manually symlink the GUI frameworks into the environment.
```
"Alternatively <comma> you can"

mdboom commented on doc/faq/virtualenv_faq.rst:
```diff
@@ -0,0 +1,106 @@
+this can be toggled with the ``toggleglobalsitepackages`` command.
+
+Alternatively you can manually symlink the GUI frameworks into the environment.
+I.e. to use PyQt5 you should symlink ``PyQt5`` and ``sip`` from your system
```
I.e. <comma> to use ``PyQt5`` <comma>

mdboom commented on doc/faq/virtualenv_faq.rst:
```diff
@@ -0,0 +1,106 @@
+OSX
+===
+
+On OSX two different types of Python Builds exists: a regular build and a
```
On OSX **comma** two different types of Python builds **exist**

mdboom commented on doc/faq/virtualenv_faq.rst:
```diff
@@ -0,0 +1,106 @@
+
+On OSX two different types of Python Builds exists: a regular build and a
+framework build. In order to interact correctly with OSX through some
+gui frameworks you need a framework build of Python.
```
gui -> GUI

mdboom commented on doc/faq/virtualenv_faq.rst:
```diff
@@ -0,0 +1,106 @@
+framework build even if created from a framework build of Python. From
+Matplotlib 1.5 onwards the ``macosx`` backend checks that a framework build is
+available and fails if a non framework build is found.
+WX has a similar check build-in.
```
build-in -> built in

mdboom commented on doc/faq/virtualenv_faq.rst:
```diff
@@ -0,0 +1,106 @@
+<https://github.com/pypa/virtualenv/issues/54>`__ and `here
+<https://github.com/pypa/virtualenv/issues/609>`__
+
+Until this is fixed a workaround is needed. The best known workaround,
```
Until this is fixed **comma** a workaround

mdboom commented on doc/faq/virtualenv_faq.rst:
```diff
@@ -0,0 +1,106 @@
+Until this is fixed a workaround is needed. The best known workaround,
+borrowed  from the `WX wiki
+<http://wiki.wxpython.org/wxPythonVirtualenvOnMac>`_, is to  use the non
+virtualenv python along with the PYTHONHOME environmental variable.  This can be
```
environmental -> environment

mdboom commented:
This is good.  Reminds me why I sometimes want to throw Macs across the room, but I think this will be very helpful.

matthew-brett commented:
If conda envs are affected the same way, should note that here?

I use the OSX backend very often in light virtualenvs, so any option that allows me to take responsibility for the risk of OSX misbehavior in a virtualenv would be very good.

jenshnielsen commented:
@matthew-brett It seems like conda envs work correctly so I have updated the document to reflect this

jenshnielsen commented:
@matthew-brett I don't have any strong opinion if this should be a warning or an error but this PR is only about the documentation

jenshnielsen commented:
Anything else I need to do here? I think the question of warning vs error should be handled outside this PR

mdboom commented:
Other than my one inline comment about linking to virtualenv docs, I think this is good to go.  I agree, let's keep this documentation-only, and think about warnings vs. exceptions later.

jenshnielsen commented:
I have added the links to the virtualenv docs and a link to an alternative workaround

mdboom commented:
Thanks!
